File: apps/worker/src/ai-provider.ts
import { createHash } from "node:crypto";

export interface AiProviderRequest {
  readonly sessionId: string;
  readonly babyId: string | null;
  readonly message: string;
  readonly attachmentIds: ReadonlyArray<string>;
  readonly onTextDelta?: (delta: string) => Promise<void>;
}

export interface AiProviderAction {
  readonly actionId: string;
  readonly entityType: string;
  readonly operation: string;
  readonly summary: string;
  readonly payload: Record<string, unknown>;
}

export interface AiProviderResult {
  readonly text: string;
  readonly actions: ReadonlyArray<AiProviderAction>;
  readonly usage?: Record<string, unknown>;
}

export interface AiProvider {
  readonly name: string;
  generate(request: AiProviderRequest): Promise<AiProviderResult>;
}

export class AiProviderError extends Error {
  readonly code: string;
  readonly retryable: boolean;
  readonly statusCode: number;

  constructor(
    message: string,
    code = "AI_PROVIDER_ERROR",
    options: { readonly retryable?: boolean; readonly statusCode?: number } = {},
  ) {
    super(message);
    this.name = "AiProviderError";
    this.code = code;
    this.retryable = options.retryable ?? false;
    this.statusCode = options.statusCode ?? 502;
  }
}

export interface AiProviderConfig {
  readonly mode: "unconfigured" | "fixture" | "openai-compatible";
  readonly baseUrl: string | null;
  readonly apiKey: string | null;
  readonly model: string | null;
  readonly timeoutMs: number;
  readonly fixtureResponse: string | null;
}

const DEFAULT_TIMEOUT_MS = 10_000;
const MAX_TIMEOUT_MS = 180_000;

function parseTimeout(raw: string | undefined): number {
  if (!raw) return DEFAULT_TIMEOUT_MS;
  const value = Number(raw);
  if (!Number.isFinite(value) || value < 100 || value > MAX_TIMEOUT_MS) {
    throw new AiProviderError(
      `GROWDESK_AI_TIMEOUT_MS must be between 100 and ${MAX_TIMEOUT_MS}`,
      "AI_PROVIDER_CONFIG_INVALID",
      { statusCode: 500 },
    );
  }
  return Math.floor(value);
}

function normalizeBaseUrl(raw: string): string {
  let url: URL;
  try {
    url = new URL(raw);
  } catch {
    throw new AiProviderError("GROWDESK_AI_BASE_URL must be a valid URL", "AI_PROVIDER_CONFIG_INVALID", {
      statusCode: 500,
    });
  }
  if (url.protocol !== "https:" && url.protocol !== "http:") {
    throw new AiProviderError(
      "GROWDESK_AI_BASE_URL must use http or https",
      "AI_PROVIDER_CONFIG_INVALID",
      { statusCode: 500 },
    );
  }
  const pathname = url.pathname.replace(/\/+$/, "");
  url.pathname = pathname.endsWith("/chat/completions")
    ? pathname
    : `${pathname}/chat/completions`;
  url.search = "";
  url.hash = "";
  return url.toString();
}

/** Resolve configuration without opening a socket or contacting a provider. */
export function resolveAiProviderConfig(env: NodeJS.ProcessEnv = process.env): AiProviderConfig {
  const mode = (env.GROWDESK_AI_PROVIDER ?? "").trim().toLowerCase();
  const timeoutMs = parseTimeout(env.GROWDESK_AI_TIMEOUT_MS);

  if (mode === "") {
    return { mode: "unconfigured", baseUrl: null, apiKey: null, model: null, timeoutMs, fixtureResponse: null };
  }
  if (mode === "fixture") {
    return {
      mode,
      baseUrl: null,
      apiKey: null,
      model: env.GROWDESK_AI_MODEL?.trim() || "fixture",
      timeoutMs,
      fixtureResponse: env.GROWDESK_AI_FIXTURE_RESPONSE ?? env.GROWDESK_AI_FIXTURE_TEXT ?? null,
    };
  }
  if (mode !== "openai-compatible" && mode !== "openai" && mode !== "compat") {
    throw new AiProviderError(`Unsupported GROWDESK_AI_PROVIDER '${mode}'`, "AI_PROVIDER_CONFIG_INVALID", {
      statusCode: 500,
    });
  }

  const baseRaw = env.GROWDESK_AI_BASE_URL?.trim();
  const apiKey = (env.GROWDESK_AI_API_KEY ?? env.AI_API_KEY ?? env.OPENAI_API_KEY ?? "").trim();
  if (!baseRaw || !apiKey) {
    return {
      mode: "unconfigured",
      baseUrl: baseRaw ? normalizeBaseUrl(baseRaw) : null,
      apiKey: apiKey || null,
      model: env.GROWDESK_AI_MODEL?.trim() || null,
      timeoutMs,
      fixtureResponse: null,
    };
  }
  return {
    mode: "openai-compatible",
    baseUrl: normalizeBaseUrl(baseRaw),
    apiKey,
    model: env.GROWDESK_AI_MODEL?.trim() || "gpt-4o-mini",
    timeoutMs,
    fixtureResponse: null,
  };
}

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  return value as Record<string, unknown>;
}

function stripJsonFence(value: string): string {
  const trimmed = value.trim();
  const fenced = trimmed.match(/^```(?:json)?\s*([\s\S]*?)\s*```$/i);
  return fenced?.[1]?.trim() ?? trimmed;
}

function parseActions(value: unknown): AiProviderAction[] {
  if (value === undefined || value === null) return [];
  if (!Array.isArray(value)) {
    throw new AiProviderError("Provider actions must be an array", "AI_PROVIDER_INVALID_RESPONSE");
  }
  return value.map((candidate, index) => {
    const action = asRecord(candidate);
    const actionId = typeof action?.actionId === "string" ? action.actionId : "";
    const entityType = typeof action?.entityType === "string" ? action.entityType : "";
    const operation = typeof action?.operation === "string" ? action.operation : "";
    const summary = typeof action?.summary === "string" ? action.summary : "";
    const payload = asRecord(action?.payload);
    if (!actionId || !entityType || !operation || !summary || !payload) {
      throw new AiProviderError(
        `Provider action at index ${index} is missing actionId, entityType, operation, summary, or payload`,
        "AI_PROVIDER_INVALID_RESPONSE",
      );
    }
    return { actionId, entityType, operation, summary, payload };
  });
}

function parseAssistantContent(content: string): { text: string; actions: AiProviderAction[] } {
  const normalized = stripJsonFence(content);
  try {
    const parsed = JSON.parse(normalized) as unknown;
    const record = asRecord(parsed);
    if (!record) return { text: content, actions: [] };
    return {
      text: typeof record.text === "string" ? record.text : "",
      actions: parseActions(record.actions ?? record.proposedActions),
    };
  } catch (error) {
    if (error instanceof AiProviderError) throw error;
    return { text: content, actions: [] };
  }
}

async function emitText(request: AiProviderRequest, delta: string): Promise<void> {
  if (delta && request.onTextDelta) await request.onTextDelta(delta);
}

function parseFixtureResponse(raw: string): { text: string; actions: AiProviderAction[]; usage?: Record<string, unknown> } {
  const parsedContent = parseAssistantContent(raw);
  let value: unknown = null;
  try {
    value = JSON.parse(raw);
  } catch {
    // Plain fixture text is valid and intentionally has no actions.
  }
  const record = asRecord(value);
  if (!record) return parsedContent;
  const text = typeof record.text === "string" ? record.text : parsedContent.text;
  const actions = record.actions === undefined ? parsedContent.actions : parseActions(record.actions);
  return { text, actions, usage: asRecord(record.usage) ?? undefined };
}

class FixtureAiProvider implements AiProvider {
  readonly name = "fixture";

  constructor(private readonly config: AiProviderConfig) {}

  async generate(request: AiProviderRequest): Promise<AiProviderResult> {
    if (!this.config.fixtureResponse) {
      throw new AiProviderError(
        "GROWDESK_AI_FIXTURE_RESPONSE or GROWDESK_AI_FIXTURE_TEXT is required when fixture provider is selected",
        "AI_PROVIDER_NOT_CONFIGURED",
        { statusCode: 503 },
      );
    }
    const result = parseFixtureResponse(this.config.fixtureResponse);
    await emitText(request, result.text);
    return result;
  }
}

interface ProviderStreamState {
  text: string;
  toolArguments: string;
  usage?: Record<string, unknown>;
}

async function parseStreamData(
  data: string,
  state: ProviderStreamState,
  request: AiProviderRequest,
): Promise<void> {
  if (data === "[DONE]") return;
  let parsed: unknown;
  try {
    parsed = JSON.parse(data);
  } catch {
    throw new AiProviderError("Provider returned malformed streaming JSON", "AI_PROVIDER_INVALID_RESPONSE");
  }
  const root = asRecord(parsed);
  const choices = Array.isArray(root?.choices) ? root.choices : [];
  const choice = asRecord(choices[0]);
  const delta = asRecord(choice?.delta);
  const content = typeof delta?.content === "string" ? delta.content : "";
  if (content) state.text += content;
  const toolCalls = Array.isArray(delta?.tool_calls) ? delta.tool_calls : [];
  for (const toolCall of toolCalls) {
    const fn = asRecord(asRecord(toolCall)?.function);
    if (typeof fn?.arguments === "string") state.toolArguments += fn.arguments;
  }
  const usage = asRecord(root?.usage);
  if (usage) state.usage = usage;
  await emitText(request, content);
}

async function readResponseBody(response: Response): Promise<string> {
  if (!response.body) return "";
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let output = "";
  while (true) {
    const chunk = await reader.read();
    if (chunk.done) break;
    output += decoder.decode(chunk.value, { stream: true });
  }
  return output + decoder.decode();
}

class OpenAiCompatibleProvider implements AiProvider {
  readonly name = "openai-compatible";

  constructor(private readonly config: AiProviderConfig) {}

  async generate(request: AiProviderRequest): Promise<AiProviderResult> {
    if (!this.config.baseUrl || !this.config.apiKey || !this.config.model) {
      throw new AiProviderError(
        "GROWDESK_AI_BASE_URL, GROWDESK_AI_API_KEY, and GROWDESK_AI_MODEL are required",
        "AI_PROVIDER_NOT_CONFIGURED",
        { statusCode: 503 },
      );
    }
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.config.timeoutMs);
    try {
      const response = await fetch(this.config.baseUrl, {
        method: "POST",
        headers: { "content-type": "application/json", authorization: `Bearer ${this.config.apiKey}` },
        body: JSON.stringify({
          model: this.config.model,
          stream: true,
          messages: [
            {
              role: "system",
              content:
                "You are GrowDesk's private baby-care assistant. Return JSON with a text string and an actions array. Use an empty actions array for read-only answers.",
            },
            { role: "user", content: request.message },
          ],
          metadata: { sessionId: request.sessionId, babyId: request.babyId, attachmentIds: request.attachmentIds },
        }),
        signal: controller.signal,
      });
      if (!response.ok) {
        const retryable = response.status === 408 || response.status === 409 || response.status === 429 || response.status >= 500;
        throw new AiProviderError(
          `AI provider returned HTTP ${response.status}`,
          response.status === 401 || response.status === 403 ? "AI_PROVIDER_AUTH_FAILED" : "AI_PROVIDER_HTTP_ERROR",
          { retryable, statusCode: response.status === 429 ? 503 : 502 },
        );
      }
      const contentType = response.headers.get("content-type")?.toLowerCase() ?? "";
      const raw = await readResponseBody(response);
      if (contentType.includes("text/event-stream")) {
        const state: ProviderStreamState = { text: "", toolArguments: "" };
        const normalized = raw.replace(/\r\n/g, "\n");
        for (const frame of normalized.split("\n\n")) {
          const dataLines = frame
            .split("\n")
            .filter((line) => line.startsWith("data:"))
            .map((line) => line.slice(5).trimStart());
          if (dataLines.length > 0) await parseStreamData(dataLines.join("\n"), state, request);
        }
        let actions: AiProviderAction[] = [];
        if (state.toolArguments) {
          let toolPayload: unknown;
          try {
            toolPayload = JSON.parse(state.toolArguments);
          } catch {
            throw new AiProviderError("Provider returned malformed tool arguments", "AI_PROVIDER_INVALID_RESPONSE");
          }
          const toolRecord = asRecord(toolPayload);
          actions = parseActions(toolRecord?.actions ?? toolPayload);
        }
        const parsed = parseAssistantContent(state.text);
        return { text: parsed.text || state.text, actions: actions.length > 0 ? actions : parsed.actions, usage: state.usage };
      }
      let payload: unknown;
      try {
        payload = JSON.parse(raw);
      } catch {
        throw new AiProviderError("Provider returned malformed JSON", "AI_PROVIDER_INVALID_RESPONSE");
      }
      const root = asRecord(payload);
      const choices = Array.isArray(root?.choices) ? root.choices : [];
      const choice = asRecord(choices[0]);
      const message = asRecord(choice?.message);
      const content = typeof message?.content === "string" ? message.content : "";
      const parsed = parseAssistantContent(content);
      await emitText(request, parsed.text || content);
      return { text: parsed.text || content, actions: parsed.actions, usage: asRecord(root?.usage) ?? undefined };
    } catch (error) {
      if (error instanceof AiProviderError) throw error;
      if (error instanceof Error && error.name === "AbortError") {
        throw new AiProviderError("AI provider request timed out", "AI_PROVIDER_TIMEOUT", { retryable: true, statusCode: 503 });
      }
      throw new AiProviderError(error instanceof Error ? error.message : String(error), "AI_PROVIDER_NETWORK_ERROR", {
        retryable: true,
        statusCode: 503,
      });
    } finally {
      clearTimeout(timeout);
    }
  }
}

export function createAiProvider(env: NodeJS.ProcessEnv = process.env): AiProvider {
  const config = resolveAiProviderConfig(env);
  if (config.mode === "fixture") return new FixtureAiProvider(config);
  if (config.mode === "openai-compatible") return new OpenAiCompatibleProvider(config);
  return {
    name: "unconfigured",
    async generate(): Promise<AiProviderResult> {
      throw new AiProviderError(
        "An AI provider is not configured. Set GROWDESK_AI_PROVIDER=fixture with an explicit fixture response for tests, or configure the approved provider and key.",
        "AI_PROVIDER_NOT_CONFIGURED",
        { statusCode: 503 },
      );
    },
  };
}

export function planHash(actions: ReadonlyArray<AiProviderAction>): string {
  return createHash("sha256").update(JSON.stringify(actions)).digest("hex");
}

File: internal/backend/ai_provider.go
package backend

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

type nativeAIAction struct {
	ActionID string `json:"actionId"`
	EntityType string `json:"entityType"`
	Operation string `json:"operation"`
	Summary string `json:"summary"`
	Payload Object `json:"payload"`
}
type nativeAIResult struct { Text string; Actions []nativeAIAction; Usage Object }
type nativeProviderConfig struct { Mode,Endpoint,APIKey,Model,Fixture,ASRModel string; Timeout time.Duration }
type nativeProviderError struct { Code string; Retryable bool; Message string }
func(e *nativeProviderError)Error()string{return e.Message}
func providerFailure(code,message string,retryable bool)error{return &nativeProviderError{Code:code,Message:message,Retryable:retryable}}
const maxProviderResponse = 4*1024*1024

func nativeProviderConfiguration()(nativeProviderConfig,error){
	c:=nativeProviderConfig{Mode:strings.ToLower(strings.TrimSpace(os.Getenv("GROWDESK_AI_PROVIDER"))),Timeout:10*time.Second,Model:strings.TrimSpace(os.Getenv("GROWDESK_AI_MODEL")),ASRModel:strings.TrimSpace(os.Getenv("GROWDESK_ASR_MODEL"))}
	if raw:=os.Getenv("GROWDESK_AI_TIMEOUT_MS");raw!=""{n,e:=strconv.ParseInt(raw,10,64);if e!=nil||n<100||n>180000{return c,apiError(500,"AI_PROVIDER_CONFIG_INVALID","AI timeout must be between 100 and 180000 milliseconds")};c.Timeout=time.Duration(n)*time.Millisecond}
	if c.Mode=="fixture"{c.Fixture=os.Getenv("GROWDESK_AI_FIXTURE_RESPONSE");if c.Fixture==""{c.Fixture=os.Getenv("GROWDESK_AI_FIXTURE_TEXT")};if c.Fixture!=""{return c,nil}}
	if c.Mode=="openai"||c.Mode=="compat"{c.Mode="openai-compatible"}
	if c.Mode=="openai-compatible"{
		c.APIKey=strings.TrimSpace(os.Getenv("GROWDESK_AI_API_KEY"));if c.APIKey==""{c.APIKey=strings.TrimSpace(os.Getenv("AI_API_KEY"))};if c.APIKey==""{c.APIKey=strings.TrimSpace(os.Getenv("OPENAI_API_KEY"))}
		u,err:=url.Parse(strings.TrimSpace(os.Getenv("GROWDESK_AI_BASE_URL")))
		if err==nil&&u.Hostname()!=""&&(u.Scheme=="https"||u.Scheme=="http")&&u.User==nil&&u.RawQuery==""&&u.Fragment==""&&c.APIKey!=""{
			u.Path=strings.TrimRight(u.Path,"/");if !strings.HasSuffix(u.Path,"/chat/completions"){u.Path+="/chat/completions"};c.Endpoint=u.String()
			if c.Model==""{c.Model="gpt-4o-mini"};if c.ASRModel==""{c.ASRModel="whisper-1"};return c,nil
		}
	}
	return c,apiError(503,"AI_PROVIDER_NOT_CONFIGURED","Configure an approved AI provider, or an explicit isolated-test fixture")
}

func providerHTTPClient(timeout time.Duration)*http.Client{
	return &http.Client{Timeout:timeout,Transport:&http.Transport{Proxy:http.ProxyFromEnvironment,DialContext:(&net.Dialer{Timeout:5*time.Second,KeepAlive:30*time.Second}).DialContext,TLSHandshakeTimeout:5*time.Second,ResponseHeaderTimeout:timeout,IdleConnTimeout:30*time.Second,MaxIdleConns:8,MaxIdleConnsPerHost:4,MaxConnsPerHost:8},CheckRedirect:func(*http.Request,[]*http.Request)error{return http.ErrUseLastResponse}}
}

var actionUUID=regexp.MustCompile(`(?i)^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`)
func parseNativeAssistant(raw string)(nativeAIResult,error){
	out:=nativeAIResult{Text:raw,Actions:[]nativeAIAction{}}
	normalized:=strings.TrimSpace(raw)
	if strings.HasPrefix(normalized,"```")&&strings.HasSuffix(normalized,"```"){normalized=strings.TrimSpace(strings.TrimSuffix(strings.TrimPrefix(normalized,"```"),"```"));normalized=strings.TrimSpace(strings.TrimPrefix(normalized,"json"))}
	var object Object
	if err:=decodeJSON([]byte(normalized),&object);err!=nil||object==nil{return out,nil}
	out.Text=text(object["text"]);out.Usage=obj(object["usage"])
	actions:=object["actions"];if actions==nil{actions=object["proposedActions"]};if actions==nil{return out,nil}
	items,ok:=actions.([]any);if !ok||len(items)>32{return out,providerFailure("AI_PROVIDER_INVALID_RESPONSE","Provider returned an invalid action list",false)}
	seen:=map[string]bool{}
	for _,item:=range items{
		a:=obj(item);action:=nativeAIAction{ActionID:text(a["actionId"]),EntityType:text(a["entityType"]),Operation:text(a["operation"]),Summary:text(a["summary"]),Payload:obj(a["payload"])}
		if !actionUUID.MatchString(action.ActionID)||seen[action.ActionID]||action.EntityType==""||action.Operation==""||action.Summary==""||action.Payload==nil{return out,providerFailure("AI_PROVIDER_INVALID_RESPONSE","Provider returned an invalid or duplicate action",false)}
		seen[action.ActionID]=true;out.Actions=append(out.Actions,action)
	}
	return out,nil
}

func callNativeAI(ctx context.Context,c nativeProviderConfig,sessionID,babyID,message string,attachments []string,delta func(string)error)(nativeAIResult,error){
	if len(message)>256*1024||len(attachments)>16{return nativeAIResult{},providerFailure("AI_INPUT_TOO_LARGE","AI input exceeds the configured budget",false)}
	if c.Mode=="fixture"{
		select{case<-ctx.Done():return nativeAIResult{},ctx.Err();default:}
		if c.Fixture==""{return nativeAIResult{},providerFailure("AI_PROVIDER_NOT_CONFIGURED","Explicit fixture content is required",false)}
		result,err:=parseNativeAssistant(c.Fixture);if err==nil&&delta!=nil&&result.Text!=""{err=delta(result.Text)};return result,err
	}
	if c.Mode!="openai-compatible"||c.Endpoint==""||c.APIKey==""{return nativeAIResult{},providerFailure("AI_PROVIDER_NOT_CONFIGURED","AI provider is unavailable",false)}
	requestBody:=Object{"model":c.Model,"stream":true,"messages":[]Object{{"role":"system","content":"You are GrowDesk's private baby-care assistant. Return JSON with a text string and an actions array. Use an empty actions array for read-only answers. Actions are proposals only; never claim a write has occurred. The server supplies the authorization scope."},{"role":"user","content":message}},"metadata":Object{"sessionId":sessionID,"babyId":babyID,"attachmentIds":attachments}}
	raw,err:=jsonBytes(requestBody);if err!=nil{return nativeAIResult{},err}
	ctx,cancel:=context.WithTimeout(ctx,c.Timeout);defer cancel()
	req,err:=http.NewRequestWithContext(ctx,http.MethodPost,c.Endpoint,bytes.NewReader(raw));if err!=nil{return nativeAIResult{},providerFailure("AI_PROVIDER_CONFIG_INVALID","Invalid AI endpoint",false)}
	req.Header.Set("Content-Type","application/json");req.Header.Set("Authorization","Bearer "+c.APIKey)
	client:=providerHTTPClient(c.Timeout);defer client.CloseIdleConnections()
	response,err:=client.Do(req)
	if err!=nil{if errors.Is(ctx.Err(),context.DeadlineExceeded){return nativeAIResult{},providerFailure("AI_PROVIDER_TIMEOUT","AI provider request timed out",true)};if ctx.Err()!=nil{return nativeAIResult{},ctx.Err()};return nativeAIResult{},providerFailure("AI_PROVIDER_NETWORK_ERROR","AI provider connection failed",true)}
	defer response.Body.Close()
	if response.StatusCode<200||response.StatusCode>=300{code:="AI_PROVIDER_HTTP_ERROR";if response.StatusCode==401||response.StatusCode==403{code="AI_PROVIDER_AUTH_FAILED"};return nativeAIResult{},providerFailure(code,fmt.Sprintf("AI provider returned HTTP %d",response.StatusCode),response.StatusCode==408||response.StatusCode==409||response.StatusCode==429||response.StatusCode>=500)}
	if strings.Contains(strings.ToLower(response.Header.Get("Content-Type")),"text/event-stream"){return readNativeAIStream(response.Body,delta)}
	body,err:=io.ReadAll(io.LimitReader(response.Body,maxProviderResponse+1));if err!=nil{return nativeAIResult{},err};if len(body)>maxProviderResponse{return nativeAIResult{},providerFailure("AI_PROVIDER_INVALID_RESPONSE","AI response exceeds the size budget",false)}
	var object Object;if err=decodeJSON(body,&object);err!=nil{return nativeAIResult{},providerFailure("AI_PROVIDER_INVALID_RESPONSE","AI provider returned malformed JSON",false)}
	choices,ok:=object["choices"].([]any);if !ok||len(choices)==0{return nativeAIResult{},providerFailure("AI_PROVIDER_INVALID_RESPONSE","AI provider returned no choices",false)}
	content:=text(obj(obj(choices[0])["message"])["content"])
	result,err:=parseNativeAssistant(content);if err!=nil{return result,err};if result.Text==""{result.Text=content};result.Usage=obj(object["usage"])
	if delta!=nil&&result.Text!=""{err=delta(result.Text)};return result,err
}

// Stream chunks are processed as they arrive. The byte, line and aggregate
// budgets bound memory; cancellation closes the HTTP body through its context.
func readNativeAIStream(reader io.Reader,delta func(string)error)(nativeAIResult,error){
	limited:=&io.LimitedReader{R:reader,N:maxProviderResponse+1}
	scanner:=bufio.NewScanner(limited);scanner.Buffer(make([]byte,4096),256*1024)
	var frame []string
	var content,arguments strings.Builder
	var usage Object
	done:=false
	consume:=func()error{
		if len(frame)==0{return nil};data:=strings.Join(frame,"\n");frame=nil
		if data=="[DONE]"{done=true;return nil}
		var object Object;if err:=decodeJSON([]byte(data),&object);err!=nil{return providerFailure("AI_PROVIDER_INVALID_RESPONSE","Malformed AI streaming JSON",false)}
		if value:=obj(object["usage"]);value!=nil{usage=value}
		choices,_:=object["choices"].([]any);if len(choices)==0{return nil}
		chunk:=obj(obj(choices[0])["delta"]);textDelta:=text(chunk["content"])
		content.WriteString(textDelta)
		if calls,ok:=chunk["tool_calls"].([]any);ok{for _,call:=range calls{arguments.WriteString(text(obj(obj(call)["function"])["arguments"]))}}
		if content.Len()+arguments.Len()>maxProviderResponse{return providerFailure("AI_PROVIDER_INVALID_RESPONSE","AI output exceeds the size budget",false)}
		if delta!=nil&&textDelta!=""{return delta(textDelta)};return nil
	}
	for scanner.Scan(){line:=scanner.Text();if line==""{if err:=consume();err!=nil{return nativeAIResult{},err};if done{break}}else if strings.HasPrefix(line,"data:"){frame=append(frame,strings.TrimLeft(strings.TrimPrefix(line,"data:")," "))}}
	if err:=scanner.Err();err!=nil{return nativeAIResult{},providerFailure("AI_PROVIDER_INVALID_RESPONSE","AI stream exceeded its line budget or was interrupted",false)}
	if limited.N<=0{return nativeAIResult{},providerFailure("AI_PROVIDER_INVALID_RESPONSE","AI stream exceeds the byte budget",false)}
	if err:=consume();err!=nil{return nativeAIResult{},err}
	if !done{return nativeAIResult{},providerFailure("AI_PROVIDER_INVALID_RESPONSE","AI stream ended without its completion marker",true)}
	result,err:=parseNativeAssistant(content.String());if err!=nil{return result,err}
	if result.Text==""{result.Text=content.String()};result.Usage=usage
	if arguments.Len()>0{
		var tool any;if err:=decodeJSON([]byte(arguments.String()),&tool);err!=nil{return result,providerFailure("AI_PROVIDER_INVALID_RESPONSE","Malformed AI tool arguments",false)}
		if obj(tool)==nil{tool=Object{"actions":tool}}
		raw,err:=jsonText(tool);if err!=nil{return result,err};parsed,err:=parseNativeAssistant(raw);if err!=nil{return result,err};result.Actions=parsed.Actions
	}
	return result,nil
}
